src/core/ddsi/src/ddsi_security_omg.c

```diff
@@ -2334,73 +2441,83 @@ static bool q_omg_security_register_remote_reader_match(struct proxy_reader *prd
-       * Simulate that we already got them. */
-      match->matched = true;
-      GVTRACE(" volatile secure writer: proxypp_crypto=%"PRId64" wr_crypto=%"PRId64" prd_crypto=%"PRId64"\n", proxypp->sec_attr->crypto_handle, wr->sec_attr->crypto_handle, match->crypto_handle);
+      GVTRACE(" match_remote_reader: proxypp_crypto=%"PRId64" wr_crypto=%"PRId64" prd_crypto=%"PRId64"\n", proxypp->sec_attr->crypto_handle, wr->sec_attr->crypto_handle, *crypto_handle);
+      send_tokens = (wr->e.guid.entityid.u != NN_ENTITYID_P2P_BUILTIN_PARTICIPANT_VOLATILE_SECURE_WRITER);
+      allowed = true;
     }
     else
+      EXCEPTION_ERROR(sc, &exception, "Failed to register remote reader "PGUIDFMT" with writer "PGUIDFMT, PGUID(prd->e.guid), PGUID(wr->e.guid));
+  }
+  else
+  {
+    struct pending_match *pending_match = find_or_create_pending_entity_match(&sc->security_matches, EK_PROXY_READER, &prd->e.guid, &wr->e.guid, 0, NULL);
+
+    /* Generate writer crypto info. */
+    if (pending_match->crypto_handle == 0)
     {
-      send_tokens = true;
-      if (match->tokens)
+      *crypto_handle = sc->crypto_context->crypto_key_factory->register_matched_remote_datareader(
+          sc->crypto_context->crypto_key_factory, wr->sec_attr->crypto_handle, proxypp->sec_attr->crypto_handle, proxypp_match->shared_secret, relay_only, &exception);
+      if (*crypto_handle == 0)
+        EXCEPTION_ERROR(sc, &exception, "Failed to register remote reader "PGUIDFMT" with writer "PGUIDFMT, PGUID(prd->e.guid), PGUID(wr->e.guid));
+      else
       {
-        if (sc->crypto_context->crypto_key_exchange->set_remote_datareader_crypto_tokens(
-            sc->crypto_context->crypto_key_exchange, wr->sec_attr->crypto_handle, match->crypto_handle, match->tokens, &exception))
+        pending_match->crypto_handle = *crypto_handle;
+        send_tokens = true;
+        GVTRACE(" register_remote_reader_match: proxypp_crypto=%"PRId64" wr_crypto=%"PRId64" prd_crypto=%"PRId64"\n", proxypp->sec_attr->crypto_handle, wr->sec_attr->crypto_handle, *crypto_handle);
+        if (pending_match->tokens)
         {
-          match->matched = true;
-          DDS_Security_DataHolderSeq_free(match->tokens);
-          match->tokens = NULL;;
-          GVTRACE("match_remote_reader "PGUIDFMT" with writer "PGUIDFMT": tokens available\n", PGUID(prd->e.guid), PGUID(wr->e.guid));
+          if (sc->crypto_context->crypto_key_exchange->set_remote_datareader_crypto_tokens(
+              sc->crypto_context->crypto_key_exchange, wr->sec_attr->crypto_handle, *crypto_handle, pending_match->tokens, &exception))
+          {
+            GVTRACE(" match_remote_reader "PGUIDFMT" with writer "PGUIDFMT": tokens available\n", PGUID(prd->e.guid), PGUID(wr->e.guid));
+            allowed = true;
+          }
+          else
+            EXCEPTION_ERROR(sc, &exception, "Failed to set remote reader crypto tokens "PGUIDFMT" --> "PGUIDFMT, PGUID(prd->e.guid), PGUID(wr->e.guid));
```

Comment:
Is ``remove_pending_match`` really the correct error handling if this error occurs? I'm just asking: it might be, but it isn't obvious.

Reply:
The problem is what to do when setting the crypto tokens fails. When that happens there is no way to recover from this. The situation will not change. Thus for the moment only a warning could be issued and the pending tokens be freed.